src/file_io.py

```diff
@@ -107,6 +109,18 @@ def save_item(file, item_):
             wrapper.write(fmt % values + "\n")
 
 
+def save_project(self, require_dialog=False, close=False):
+    if self.get_data().project_uri != "" and not require_dialog:
+        file_uri = self.get_data().project_uri
+        file = Gio.File.new_for_uri(file_uri)
+        write_json(file, self.get_data().to_project_dict(), False)
+        self.get_data().props.unsaved = False
+        if close:
```

Comment:
This is not super pretty. The issue is that Python doesn't wait until a response before executing the next line. So in the following dialog:

```
        if self.get_data().props.unsaved:
            def on_response(_dialog, response):
                if response == "discard_close":
                    self.quit()
                if response == "save_close":
                    file_io.save_project(self, close=True)

            dialog = ui.build_dialog("save_changes")
            dialog.set_transient_for(self.get_window())
            dialog.connect("response", on_response)
            dialog.present()
```
It would probably be nicest to handle the "close"  signal directly after the `if response == "save_close":` line. But if that is done, then the save dialog is shown, and immediately the application will be closed before the user even has the option to choose something. So either way, one would need to wait for a response.

I've tried to look into some solutions. I looked a bit after trying to emit a specific signal, but that didn't become much prettier. Ideally, we would maybe just halt everything when a GtkFiledialog is shown, and wait until that is resolved. But I am not 100% if that's possible, or compatible with modern GObject logic.

Either way, the way I do it here by handling this on the response signal works. The idea is that this may be cleaned in a future PR though. As I'd probably prefer to handle the close signal at the close dialog.